src/ripple/shamap/impl/SHAMap.cpp

```diff
@@ -615,68 +803,97 @@ bool SHAMap::delItem (uint256 const& id)
 
     // What gets attached to the end of the chain
     // (For now, nothing, since we deleted the leaf)
-    SHAMapHash prevHash;
     std::shared_ptr<SHAMapAbstractNode> prevNode;
 
     while (!stack.empty ())
     {
-        auto node = std::dynamic_pointer_cast<SHAMapInnerNode>(stack.top ().first);
-        SHAMapNodeID nodeID = stack.top ().second;
-        stack.pop ();
-
-        assert (node);
+        auto node = std::static_pointer_cast<SHAMapInnerNode>(stack.top().first);
+        SHAMapNodeID nodeID = stack.top().second;
+        stack.pop();
 
         node = unshareNode(std::move(node), nodeID);
-        node->setChild (nodeID.selectBranch (id), prevNode);
+        node->setChild(nodeID.selectBranch(id), prevNode);
 
         if (!nodeID.isRoot ())
         {
             // we may have made this a node with 1 or 0 children
             // And, if so, we need to remove this branch
-            int bc = node->getBranchCount ();
-
-            if (bc == 0)
+            int bc = node->getBranchCount();
+            if (is_v2())
             {
-                // no children below this branch
-                prevHash.zero();
-                prevNode.reset ();
-            }
-            else if (bc == 1)
-            {
-                // If there's only one item, pull up on the thread
-                std::shared_ptr<SHAMapItem const> item = onlyBelow (node.get ());
-
-                if (item)
+                assert(bc != 0);
+                if (bc == 1)
                 {
                     for (int i = 0; i < 16; ++i)
                     {
                         if (!node->isEmptyBranch (i))
                         {
-                            node->setChild (i, nullptr);
+                            prevNode = descendThrow(node, i);
                             break;
                         }
                     }
-                    prevNode = std::make_shared<SHAMapTreeNode>(item, type, node->getSeq());
-                    prevHash = prevNode->getNodeHash();
                 }
-                else
+                else  // bc >= 2
                 {
-                    prevHash = node->getNodeHash ();
-                    prevNode = std::move (node);
+                    // This node is now the end of the branch
+                    prevNode = std::move(node);
                 }
             }
             else
             {
-                // This node is now the end of the branch
-                prevHash = node->getNodeHash ();
-                prevNode = std::move (node);
+                if (bc == 0)
+                {
+                    // no children below this branch
+                    prevNode.reset ();
+                }
+                else if (bc == 1)
+                {
+                    // If there's only one item, pull up on the thread
+                    std::shared_ptr<SHAMapItem const> item = onlyBelow (node.get ());
```

Comment:
Nope.  `onlyBelow` is returning a reference to an already existing `shared_ptr`.  It never creates a new one.